File: app/javascript/controllers/rewards_controller.js
import { Controller } from "stimulus";
import Rails from "@rails/ujs";

export default class extends Controller {
  static targets = ["rewardSource", "result", "errors"];

  calculate() {
    const f = this.rewardSourceTarget.files[0];
    const resultTarget = this.resultTarget;
    const errorsTarget = this.errorsTarget;
    if (f) {
      const send = this.sendData;
      const reader = new FileReader();
      reader.onload = function (e) {
        // binary data
        send(
          { reward: { binary_data: reader.result } },
          resultTarget,
          errorsTarget
        );
      };
      reader.readAsBinaryString(f);
    }
  }

  sendData(data, resultTarget, errorsTarget) {
    Rails.ajax({
      type: "POST",
      url: "/rewards",
      beforeSend(xhr, options) {
        xhr.setRequestHeader("Content-Type", "application/json; charset=UTF-8");
        options.data = JSON.stringify(data);
        return true;
      },
      data: data,
      success: (response) => {
        errorsTarget.innerHTML = "";
        resultTarget.innerHTML = `<h4>Result</h4> ${JSON.stringify(response)}`;
      },
      error: (error) => {
        resultTarget.innerHTML = "";
        errorsTarget.innerHTML = `<div class="alert alert-danger">Issue noticed in data file</div>`;
      },
    });
  }
}

File: app/controllers/rewards_controller.rb
class RewardsController < ApplicationController
  skip_before_action :verify_authenticity_token

  def calculate
    service = Rewards::Main.new(reward_params[:binary_data])
    respond_to do |format|
      if service.invalid?
        format.json { render json: { errors: service.errors }, status: :unprocessable_entity }
      end

      format.json { render json: service.calculate, status: :ok }
    end
  end

  private

  def reward_params
    params.require(:reward).permit(:binary_data)
  end
end
